Gems/Atom/RHI/Vulkan/Code/Source/RHI/RenderPass.cpp

```diff
@@ -16,6 +16,387 @@ namespace AZ
+                Preserve, // An attachment that is not being used by the subpass but that it will be preserved.
+                ShadingRate, // An attachment that is used for specifiying shading rate.
+                Count
+            };
+            static const uint32_t AttachmentTypeCount = static_cast<uint32_t>(AttachmentType::Count);
+
+            using RenderPassResult = AZStd::tuple<VkResult, VkRenderPass>;
+
+            // Helper class to build the Vulkan native renderpass.
+            // If available at runtime, we prefer to use the Renderpass2 extension (need it for other features like shading rate attachments).
+            // This extension has different structures for declaring the attachments, subpasses, dependencies, etc. Although the logic is almost the same
+            // as the standard declaration of a renderpass, the extension has members that do not exist in the
+            // standard way. Because of this, a template approach is used to only use those members is they exist
+            // in the structure.
+            template<
+                typename VkRenderPassCreateInfoTraits,
+                typename VkAttachmentDescriptionTraits,
+                typename VkAttachmentReferenceTraits,
+                typename VkSubpassDescriptionTraits,
+                typename VkSubpassDependencyTraits>
+            struct VkRenderpassBuilder
+            {
+            public:
+
+                VkRenderpassBuilder(const Device& device)
+                    : m_device(device)
+                {}
+
+                // The structure for declaring the renderpass
+                using VkRenderPassCreateInfoType = typename VkRenderPassCreateInfoTraits::value_type;
+                // The structyure used for declaring attachments
+                using VkAttachmentDescriptionType = typename VkAttachmentDescriptionTraits::value_type;
+                // The structure used for referencing an attachment
+                using VkAttachmentReferenceType = typename VkAttachmentReferenceTraits::value_type;
+                // The structure used for describing a subpass
+                using VkSubpassDescriptionType = typename VkSubpassDescriptionTraits::value_type;
+                // The structure used for declaring a dependency
+                using VkSubpassDependencyType = typename VkSubpassDependencyTraits::value_type;
+
+                // Holds the information about a subpass
+                struct SubpassInfo
+                {
+                    AZStd::array<AZStd::vector<VkAttachmentReferenceType>, AttachmentTypeCount> m_attachmentReferences;
+                    AZStd::vector<uint32_t> m_preserveAttachments;
+                    // Used only if the pass uses fragment rate attachment.
+                    VkFragmentShadingRateAttachmentInfoKHR m_shadingRateAttachmentExtension;
+                };
+
+                RenderPassResult Build(const RenderPass::Descriptor& descriptor)
```

Comment:
Just an opinion. This class will be more readable and easy to see its interface if the functions implementations would be move outside the class.